File: android/src/main/java/com/bitmovin/player/reactnative/NetworkModule.kt
package com.bitmovin.player.reactnative

import android.util.Log
import androidx.concurrent.futures.CallbackToFutureAdapter
import androidx.concurrent.futures.CallbackToFutureAdapter.Completer
import androidx.core.os.bundleOf
import com.bitmovin.player.api.network.HttpRequest
import com.bitmovin.player.api.network.HttpRequestType
import com.bitmovin.player.api.network.HttpResponse
import com.bitmovin.player.api.network.NetworkConfig
import com.bitmovin.player.api.network.PreprocessHttpRequestCallback
import com.bitmovin.player.api.network.PreprocessHttpResponseCallback
import com.bitmovin.player.reactnative.converter.toHttpRequest
import com.bitmovin.player.reactnative.converter.toHttpResponse
import com.bitmovin.player.reactnative.converter.toJson
import com.bitmovin.player.reactnative.converter.toNetworkConfig
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Future

/**
 * Expo module for NetworkConfig management with HTTP request/response preprocessing.
 * Handles bidirectional communication between native code and JavaScript for network operations.
 */
class NetworkModule : Module() {
    /**
     * In-memory mapping from `nativeId`s to `NetworkConfig` instances.
     */
    private val networkConfigs: Registry<NetworkConfig> = mutableMapOf()
    private val preprocessHttpRequestCompleters = ConcurrentHashMap<String, Completer<HttpRequest>>()
    private val preprocessHttpResponseCompleters = ConcurrentHashMap<String, Completer<HttpResponse>>()

    override fun definition() = ModuleDefinition {
        Name("NetworkModule")

        Events("onPreprocessHttpRequest", "onPreprocessHttpResponse")

        AsyncFunction("initializeWithConfig") { nativeId: String, config: Map<String, Any?>, promise: Promise ->
            if (networkConfigs.containsKey(nativeId)) {
                promise.resolve(null)
                return@AsyncFunction
            }

            try {
                val networkConfig = config.toNetworkConfig()
                networkConfigs[nativeId] = networkConfig
                initConfigBlocks(nativeId, config)
                promise.resolve(null)
            } catch (e: Exception) {
                promise.reject("NetworkError", "Failed to initialize network config", e)
            }
        }

        AsyncFunction("destroy") { nativeId: String ->
            networkConfigs.remove(nativeId)

            // Clean up completion handlers
            preprocessHttpRequestCompleters.keys.filter { it.startsWith(nativeId) }.forEach {
                preprocessHttpRequestCompleters.remove(it)
            }
            preprocessHttpResponseCompleters.keys.filter { it.startsWith(nativeId) }.forEach {
                preprocessHttpResponseCompleters.remove(it)
            }
        }

        AsyncFunction("setPreprocessedHttpRequest") { requestId: String, request: Map<String, Any?> ->
            val completer = preprocessHttpRequestCompleters.remove(requestId)
            if (completer == null) {
                Log.e("NetworkModule", "Completer is null for requestId: $requestId, this can cause stuck network requests")
                return@AsyncFunction
            }
            completer.set(request.toHttpRequest())
        }

        AsyncFunction("setPreprocessedHttpResponse") { responseId: String, response: Map<String, Any?> ->
            preprocessHttpResponseCompleters[responseId]?.set(response.toHttpResponse())
            preprocessHttpResponseCompleters.remove(responseId)
        }
    }

    /**
     * Retrieves the NetworkConfig instance for the given nativeId.
     * This method maintains the same static access pattern as the legacy module.
     */
    fun getConfig(nativeId: String?): NetworkConfig? = nativeId?.let { networkConfigs[it] }

    private fun initConfigBlocks(nativeId: String, config: Map<String, Any?>) {
        initPreprocessHttpRequest(nativeId, config)
        initPreprocessHttpResponse(nativeId, config)
    }

    private fun initPreprocessHttpRequest(nativeId: String, networkConfigJson: Map<String, Any?>) {
        val networkConfig = getConfig(nativeId) ?: return
        if (!networkConfigJson.containsKey("preprocessHttpRequest")) return

        networkConfig.preprocessHttpRequestCallback = PreprocessHttpRequestCallback { type, request ->
            preprocessHttpRequestFromJS(nativeId, type, request)
        }
    }

    private fun initPreprocessHttpResponse(nativeId: String, networkConfigJson: Map<String, Any?>) {
        val networkConfig = getConfig(nativeId) ?: return
        if (!networkConfigJson.containsKey("preprocessHttpResponse")) return

        networkConfig.preprocessHttpResponseCallback = PreprocessHttpResponseCallback { type, response ->
            preprocessHttpResponseFromJS(nativeId, type, response)
        }
    }

    private fun preprocessHttpRequestFromJS(
        nativeId: String,
        type: HttpRequestType,
        request: HttpRequest,
    ): Future<HttpRequest> {
        val requestId = "$nativeId@${System.identityHashCode(request)}"
        val args = mapOf(
            "requestId" to requestId,
            "type" to type.toJson(),
            "request" to request.toJson(),
        )

        return CallbackToFutureAdapter.getFuture { completer ->
            preprocessHttpRequestCompleters[requestId] = completer

            // Send event to TypeScript using Expo module event system
            sendEvent(
                "onPreprocessHttpRequest",
                bundleOf(
                    "nativeId" to nativeId,
                    "requestId" to requestId,
                    "type" to type.toJson(),
                    "request" to request.toJson(),
                ),
            )

            return@getFuture "NetworkModule-preprocessHttpRequest-$requestId"
        }
    }

    private fun preprocessHttpResponseFromJS(
        nativeId: String,
        type: HttpRequestType,
        response: HttpResponse,
    ): Future<HttpResponse> {
        val responseId = "$nativeId@${System.identityHashCode(response)}"

        return CallbackToFutureAdapter.getFuture { completer ->
            preprocessHttpResponseCompleters[responseId] = completer

            // Send event to TypeScript using Expo module event system
            sendEvent(
                "onPreprocessHttpResponse",
                bundleOf(
                    "nativeId" to nativeId,
                    "responseId" to responseId,
                    "type" to type.toJson(),
                    "response" to response.toJson(),
                ),
            )

            return@getFuture "NetworkModule-preprocessHttpResponse-$responseId"
        }
    }
}

File: ios/NetworkModule.swift
import BitmovinPlayer
import BitmovinPlayerCore
import ExpoModulesCore

/**
 * Expo module for NetworkConfig management with HTTP request/response preprocessing.
 * Handles bidirectional communication between native code and JavaScript for network operations.
 */
public class NetworkModule: Module {
    /// In-memory mapping from `nativeId`s to `NetworkConfig` instances.
    private var networkConfigs: Registry<NetworkConfig> = [:]
    private var preprocessHttpRequestCompletionHandlers: Registry<(_ request: HttpRequest) -> Void> = [:]
    private var preprocessHttpResponseCompletionHandlers: Registry<(_ response: HttpResponse) -> Void> = [:]

    public func definition() -> ModuleDefinition {
        Name("NetworkModule")

        OnDestroy {
            networkConfigs.removeAll()
            preprocessHttpRequestCompletionHandlers.removeAll()
            preprocessHttpResponseCompletionHandlers.removeAll()
        }

        Events(
            "onPreprocessHttpRequest",
            "onPreprocessHttpResponse"
        )

        AsyncFunction("initializeWithConfig") { [weak self] (nativeId: NativeId, config: [String: Any]) in
            guard
                self?.retrieve(nativeId) == nil,
                let networkConfig = RCTConvert.networkConfig(config)
            else {
                return
            }
            self?.networkConfigs[nativeId] = networkConfig
            self?.initConfigBlocks(nativeId, config)
        }.runOnQueue(.main)

        AsyncFunction("destroy") { [weak self] (nativeId: NativeId) in
            self?.networkConfigs.removeValue(forKey: nativeId)

            // Clean up completion handlers
            self?.preprocessHttpRequestCompletionHandlers.keys.filter { $0.starts(with: nativeId) }.forEach {
                self?.preprocessHttpRequestCompletionHandlers.removeValue(forKey: $0)
            }
            self?.preprocessHttpResponseCompletionHandlers.keys.filter { $0.starts(with: nativeId) }.forEach {
                self?.preprocessHttpResponseCompletionHandlers.removeValue(forKey: $0)
            }
        }.runOnQueue(.main)

        AsyncFunction("setPreprocessedHttpRequest") { [weak self] (requestId: String, request: [String: Any]) in
            guard let completionHandler = self?.preprocessHttpRequestCompletionHandlers[requestId],
                  let httpRequest = RCTConvert.httpRequest(request) else {
                return
            }

            self?.preprocessHttpRequestCompletionHandlers.removeValue(forKey: requestId)
            completionHandler(httpRequest)
        }.runOnQueue(.main)

        AsyncFunction("setPreprocessedHttpResponse") { [weak self] (responseId: String, response: [String: Any]) in
            guard let completionHandler = self?.preprocessHttpResponseCompletionHandlers[responseId],
                  let httpResponse = RCTConvert.httpResponse(response) else {
                return
            }
            self?.preprocessHttpResponseCompletionHandlers.removeValue(forKey: responseId)
            completionHandler(httpResponse)
        }.runOnQueue(.main)
    }

    func retrieve(_ nativeId: NativeId) -> NetworkConfig? {
        networkConfigs[nativeId]
    }

    private func initConfigBlocks(_ nativeId: NativeId, _ config: [String: Any]) {
        initPreprocessHttpRequest(nativeId, networkConfigJson: config)
        initPreprocessHttpResponse(nativeId, networkConfigJson: config)
    }

    private func initPreprocessHttpRequest(_ nativeId: NativeId, networkConfigJson: [String: Any]) {
        guard let networkConfig = retrieve(nativeId),
              networkConfigJson["preprocessHttpRequest"] != nil else {
            return
        }
        networkConfig.preprocessHttpRequest = { [weak self] type, request, completionHandler in
            self?.preprocessHttpRequestFromJS(
                nativeId: nativeId,
                type: type,
                request: request,
                completionHandler: completionHandler
            )
        }
    }

    private func initPreprocessHttpResponse(_ nativeId: NativeId, networkConfigJson: [String: Any]) {
        guard let networkConfig = retrieve(nativeId),
              networkConfigJson["preprocessHttpResponse"] != nil else {
            return
        }

        networkConfig.preprocessHttpResponse = { [weak self] type, response, completionHandler in
            self?.preprocessHttpResponseFromJS(
                nativeId: nativeId,
                type: type,
                response: response,
                completionHandler: completionHandler
            )
        }
    }

    internal func preprocessHttpRequestFromJS(
        nativeId: NativeId,
        type: HttpRequestType,
        request: HttpRequest,
        completionHandler: @escaping (
            _ request: HttpRequest
        ) -> Void
    ) {
        let requestId = "\(nativeId)-\(UUID().uuidString)"
        preprocessHttpRequestCompletionHandlers[requestId] = completionHandler
        sendEvent("onPreprocessHttpRequest", [
            "nativeId": nativeId,
            "requestId": requestId,
            "type": RCTConvert.toJson(httpRequestType: type),
            "request": RCTConvert.toJson(httpRequest: request)
        ])
    }

    private func preprocessHttpResponseFromJS(
        nativeId: NativeId,
        type: HttpRequestType,
        response: HttpResponse,
        completionHandler: @escaping (
            _ response: HttpResponse
        ) -> Void
    ) {
        let responseId = "\(nativeId)-\(UUID().uuidString)"
        preprocessHttpResponseCompletionHandlers[responseId] = completionHandler
        sendEvent("onPreprocessHttpResponse", [
            "nativeId": nativeId,
            "responseId": responseId,
            "type": RCTConvert.toJson(httpRequestType: type),
            "response": RCTConvert.toJson(httpResponse: response)
        ])
    }
}
